File: solutions/leetcode/search-in-rotated-sorted-array.cpp
/* Problem search-in-rotated-sorted-array
 * https://www.leetcode.com/problems/search-in-rotated-sorted-array
 */

#include <vector>

using namespace std;

class Solution {
 private:
  pair<bool, vector<int>::const_iterator> rotation_point(
      vector<int>::const_iterator const first,
      vector<int>::const_iterator const last) {
    ptrdiff_t const dist = distance(first, last);
    if (dist <= 1 || *first < *prev(last)) {
      return make_pair(true, first);
    } else {
      vector<int>::const_iterator const mid = first + dist / 2;
      auto const right = rotation_point(mid, last);
      if (right.first && *prev(mid) < *mid) {
        return make_pair(false, rotation_point(first, mid).second);
      } else {
        return make_pair(false, right.second);
      }
    }
  }

 public:
  int search(vector<int> const& nums, int const target) {
    auto const mid = rotation_point(nums.cbegin(), nums.cend()).second;
    auto const left = lower_bound(nums.cbegin(), mid, target);
    if (left != nums.cend() && *left == target) return left - nums.cbegin();
    auto const right = lower_bound(mid, nums.cend(), target);
    if (right != nums.cend() && *right == target) return right - nums.cbegin();
    return -1;
  }
};

File: solutions/leetcode/search-in-rotated-sorted-array.rs
/* Problem search-in-rotated-sorted-array
 * https://www.leetcode.com/problems/search-in-rotated-sorted-array
 */

struct Solution;

impl Solution {
	fn find_pivot(nums: &[i32]) -> usize {
		if nums.len() <= 1 {
			return 0;
		}

		let mid = nums.len() / 2;
		if nums
			.get(mid)
			.zip(nums.get(mid + 1))
			.map_or(false, |(a, b)| a > b)
		{
			mid + 1
		} else {
			if *nums.last().unwrap() > nums[mid] {
				Self::find_pivot(&nums[..=mid])
			} else {
				Self::find_pivot(&nums[mid..]) + mid
			}
		}
	}

	pub fn search(nums: Vec<i32>, target: i32) -> i32 {
		let pivot = Self::find_pivot(&nums);

		let (l, r) = nums.split_at(pivot);

		l.binary_search(&target)
			.or(r.binary_search(&target).map(|index| index + pivot))
			.map(|index| index.try_into().unwrap())
			.unwrap_or(-1)
	}
}

pub fn main() {
	assert_eq!(4, Solution::search(vec![4, 5, 6, 7, 0, 1, 2], 0));
	assert_eq!(-1, Solution::search(vec![4, 5, 6, 7, 0, 1, 2], 3));
	assert_eq!(-1, Solution::search(vec![1], 0));
	assert_eq!(1, Solution::search(vec![3, 5, 1], 5));
	assert_eq!(0, Solution::search(vec![5, 1, 3], 5));
}
